@balloob, @bdraco  it seem to behaving same issue i just noticed in rfxtrx tests. It seems async_setup_component previously actually waited for the complete setup. But after the startup time improvement this doesn't hold anymore.

So for example the cast tests that fail above. The teardown of the test actually start before the whole of hass has completed it's setup. Then you get uncaught exceptions in the hass object.

```
-------------------------------------------------------------------------------------------------- Captured stderr teardown --------------------------------------------------------------------------------------------------
DEBUG:homeassistant.core:Bus:Handling <Event homeassistant_stop[L]>
DEBUG:homeassistant.setup:Dependency media_player will wait for ['http']
ERROR:homeassistant.core:Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/Users/joakim/src/hass/home-assistant/homeassistant/core.py", line 1219, in async_call
    raise ServiceNotFound(domain, service) from None
homeassistant.exceptions.ServiceNotFound: Unable to find service persistent_notification/create
ERROR:homeassistant.core:Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/Users/joakim/src/hass/home-assistant/homeassistant/core.py", line 1219, in async_call
    raise ServiceNotFound(domain, service) from None
homeassistant.exceptions.ServiceNotFound: Unable to find service persistent_notification/dismiss
INFO:homeassistant.setup:Setting up http
INFO:homeassistant.setup:Setup of domain http took 0.0 seconds.
DEBUG:homeassistant.core:Bus:Handling <Event component_loaded[L]: component=http>
INFO:homeassistant.setup:Setting up media_player
```

Which likely blocks the STOP event to propagate to all components needing it.

I'm unsure in which end we want to fix this since this likely affect many many tests. If you add a `await hass.async_block_till_done` after the async_setup_component(). the setup actually complete and the test succeed without lingering threads.

This would likely need to be added to ALL tests for ALL components using async_setup_component(). The second problem is that many tests do stuff like `assert not await async_setup_component(` which i think is no longer valid way to test for a failed setup of a component.

Any thoughts?

--- a/homeassistant/bootstrap.py
+++ b/homeassistant/bootstrap.py
@@ -138,9 +138,11 @@ async def async_setup_hass(
         with contextlib.suppress(asyncio.TimeoutError):
             async with timeout(10):
                 await hass.async_block_till_done()
+                hass.executor.shutdown(False)
 
         safe_mode = True
         old_config = hass.config
+
         hass = core.HomeAssistant()
         hass.config.skip_pip = old_config.skip_pip
         hass.config.internal_url = old_config.internal_url

--- a/tests/components/cast/test_media_player.py
+++ b/tests/components/cast/test_media_player.py
@@ -559,6 +559,7 @@ async def test_disconnect_on_stop(hass: HomeAssistantType):
 async def test_entry_setup_no_config(hass: HomeAssistantType):
     """Test setting up entry with no config.."""
     await async_setup_component(hass, "cast", {})
+    await hass.async_block_till_done()
 
     with patch(
         "homeassistant.components.cast.media_player._async_setup_platform",
@@ -574,6 +575,7 @@ async def test_entry_setup_single_config(hass: HomeAssistantType):
     await async_setup_component(
         hass, "cast", {"cast": {"media_player": {"host": "bla"}}}
     )
+    await hass.async_block_till_done()
 
     with patch(
         "homeassistant.components.cast.media_player._async_setup_platform",
@@ -589,6 +591,7 @@ async def test_entry_setup_list_config(hass: HomeAssistantType):
     await async_setup_component(
         hass, "cast", {"cast": {"media_player": [{"host": "bla"}, {"host": "blu"}]}}
     )
+    await hass.async_block_till_done()
 
     with patch(
         "homeassistant.components.cast.media_player._async_setup_platform",
@@ -605,6 +608,7 @@ async def test_entry_setup_platform_not_ready(hass: HomeAssistantType):
     await async_setup_component(
         hass, "cast", {"cast": {"media_player": {"host": "bla"}}}
     )
+    await hass.async_block_till_done()
 
     with patch(
         "homeassistant.components.cast.media_player._async_setup_platform",

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -2,6 +2,7 @@
 import asyncio
 import functools
 import logging
+import threading
 
 import pytest
 import requests_mock as _requests_mock
@@ -78,6 +79,8 @@ util.get_local_ip = lambda: "127.0.0.1"
 @pytest.fixture(autouse=True)
 def verify_cleanup():
     """Verify that the test has cleaned up resources correctly."""
+    threads_before = frozenset(threading.enumerate())
+
     yield
 
     if len(INSTANCES) >= 2:
@@ -85,6 +88,9 @@ def verify_cleanup():
         for inst in INSTANCES:
             inst.stop()
         pytest.exit(f"Detected non stopped instances ({count}), aborting test run")
+
+    threads = frozenset(threading.enumerate()) - threads_before
+    assert not threads
 
 
 @pytest.fixture

